### PR TITLE
Video tracking frontend

### PR DESCRIPTION
**Pull request by yichenshen**
This pull request logs video events on the frontend by adding a reducer to listen to certain Redux actions we are interested in logging. These events are then collated on the client, before being periodically sent back to the server (while at the same time updating the `session_end` and `last_video_time` columns for a session as a way to denote that the session is still alive.

The models are changed slightly to fit implementation requirements.
 - We don't need a `video_time_final`, the function is fulfilled with extra event types
 - `last_video_time` helps us figure out where in the video the user has reached before closing the window/disconnecting. It is to be used in conjunction with the last recorded event.

We remove events from the client once we receive a success response from the server. Only events that were sent are removed.

Race conditions are handled by a combination of `merge_events!` on the server side, which handles events being sent again before they are removed, and reducer logic on the client (reducer functions modify the Redux state in a consistent and atomic fashion, i.e. if event 3 is present, event 2 must be present either in the client, the server, or both.

### CODE REVIEW

File: app/controllers/course/video/submission/submissions_controller.rb [hunk 1]
```diff
@@ -38,4 +39,10 @@ def scroll_topic_params
   def authorize_video!
     authorize!(:attempt, @video)
   end
+
+  def create_session
+    return unless !current_course_user.staff? && @submission.course_user == current_course_user
```
Comment (weiqingtoh) [inline app/controllers/course/video/submission/submissions_controller.rb]: Can do `current_course_user.student?` instead.

Comment (yichenshen) [reply app/controllers/course/video/submission/submissions_controller.rb]: Ah ok, sure!

File: client/app/bundles/course/video/submission/actions/video.js [hunk 2]
```diff
@@ -139,3 +140,40 @@ export function seekStart() {
...
+ * Events will be removed from the state store if the API call is successful.
+ * @return {function(*, *)}
+ */
+export function sendEvents() {
+  return (dispatch, getState) => {
+    const videoState = getState().video;
+    const sessionId = videoState.sessionId;
+
+    if (sessionId === null) {
+      return;
```
Comment (weiqingtoh) [inline client/app/bundles/course/video/submission/actions/video.js]: Just to check, this is just precautionary? By right if sessionId is null there shouldn't be events being generated and pushed onto the queue?

File: client/app/bundles/course/video/submission/containers/VideoPlayer.jsx [hunk 3]
```diff
@@ -62,12 +65,24 @@ class VideoPlayer extends React.Component {
...
+    }
+  }
+
   componentWillReceiveProps(nextProps) {
     if (nextProps.forceSeek) {
       this.player.seekTo(nextProps.playerProgress);
     }
   }
 
+  componentWillUnmount() {
```
Comment (weiqingtoh) [inline client/app/bundles/course/video/submission/containers/VideoPlayer.jsx]: Just a random thought: would it be a good idea to try to send events before the component unmounts? (Probably when the browser winder closes)

File: client/app/bundles/course/video/submission/actions/video.js (hunk 2 above)
Comment (yichenshen) [reply client/app/bundles/course/video/submission/actions/video.js]: Yup, but I send a request to the server even if there are no events, so as to update the `session_end` time. If `sessionId` is null, we skip everything to not spam the server with unnecessary requests.

File: client/app/bundles/course/video/submission/containers/VideoPlayer.jsx (hunk 3 above)
Comment (yichenshen) [reply client/app/bundles/course/video/submission/containers/VideoPlayer.jsx]: That makes sense. I'll add another call to `onTick`

**Comment by weiqingtoh** [comment]
Generally, the implementation seems 👍  - have 2 concerns/questions:
 1. For the 1000 millisecond tick, @jeremyyap was suggesting we can try ActionCable (web sockets) instead, otherwise if too many people are simultaneously watching it could result in quite a few pings to the server. 
 1. For the sessions, there is a small possibility of being able to modify past sessions (if I made the right API calls)? Any thing we can do about it? The only one I can think of is to instead use a GUID (so that we don't allow random people to try running numbers), but to me that doesn't solve the problem entirely.

These 2 concerns are a little more far fetched, so just putting it out for discussion, will not need to fix in the PR.  Thoughts?

**Comment by yichenshen** [comment]
1) Yea, I was thinking about that too when I started. Websockets/ActionCable is definitely better than the current solution. My last experience with it though was that it needed some setup on the backend (in production it needs a Redis server and HTTP Upgrade needs to be allowed if there's a reverse proxy). I wasn't sure if everything can be made to work in time so I did a long poll first. I'm happy to discuss and work on switching though.

2) Yea, I did not think of this point. Another idea I have is to limit the update time. So if someone tries to update a session that was last updated say 24hrs ago, we reject the update, though I don't think that solves the problem fully either.

**Comment by weiqingtoh** [comment]
Ok - could you open a issue to track this?

One method to solve point 2 is with websockets, where a sessionId can be generated for each new websocket connection, and closed once the connection closes? I'm not sure of the implications though since I've only read up on websockets.

**Comment by yichenshen** [comment]
Alright sure.

I'm not too sure if ActionCable can tell the difference between a proper connection close and a disconnect though. If it can't, then a client which disconnects while remaining on the page might find itself unable to further append to the session.

**Comment by benleong** [comment]
Can we test this?

**Pull request merged**